### ISSUE TITLE
Zombies: async generator component causes unmounted nodes to stick around in parent

### ISSUE THREAD

**Issue by brainkim** [bikeshaving/crank#121]
Someone reported a bug in 0.1.4 where unmounted nodes stick around in certain situations. I was able to create a minimal reproduction in the test suite:

```tsx
test("unmount edge case", async () => {
	function Switch({children, active}: {children: Children, active: boolean}) {
		if (!active) {
			return null;
		}

		return children;
	}

	async function *AsyncGen(this: Context) {
		for await (const _ of this) {
			yield <span>true</span>;
		}
	}

	function *Component() {
		let toggle = true;
		while (true) {
			yield (
				<div>
					<Switch active={toggle}><AsyncGen /></Switch>
					<Switch active={!toggle}><span>false</span></Switch>
				</div>
			)

			toggle = !toggle;
		}
	}

	await renderer.render(<Component />, document.body);
	expect(document.body.innerHTML).toEqual("<div><span>true</span></div>");
	await renderer.render(<Component />, document.body);
	expect(document.body.innerHTML).toEqual("<div><span>false</span></div>");
	await renderer.render(<Component />, document.body);
	expect(document.body.innerHTML).toEqual("<div><span>true</span></div>");
});
```

So it only happens when there are sibling components where the first component’s children includes an async generator component. (It doesn’t happen when the async generator component comes second).

I’ve identified the cause of the bug as being an optimization I made where we commit separately from updating (for some reason grouping DOM updates as closely as possible seems to make things faster). Async generator components had to be commit separately because of a race condition related to resuming async generators with their value. I thought committing twice would be fine, but apparently it caused weird issues like this one.

Fix is on the way.

**Comment by ryhinchey** [comment]
Nice!  Thanks for this 😁

**Comment by brainkim** [comment]
I tried to do a quick fix which preserved the separate commit optimization, but it didn’t actually fix the bug in your realworld app @ryhinchey. It did fix the bug when you clicked home and back, but it didn’t fix it when you used the browser navigation back button, but only sometimes. This indicates some kind of race condition. I think that using async generators on multiple levels caused the same bug where the parent was committed out of order. This is scary to me so I’m backing out of the separate commit optimizations until we can understand/test this.

**Comment by brainkim** [comment]
I will make sure to test any future optimizations against the realworld app until I can fully reproduce the bug in test cases.

**Comment by brainkim** [comment]
Fixed in 0.1.6
